crates/xtask-bump-check/src/xtask.rs

```diff
@@ -0,0 +1,338 @@
+    repo: &'r git2::Repository,
+    base_commit: &git2::Commit<'r>,
+    head: &git2::Commit<'r>,
+) -> CargoResult<HashMap<&'ws str, &'ws Package>> {
+    let member_dirs = ["crates", "credential", "benches"];
+    let ws_members: HashMap<&str, &Package> =
+        HashMap::from_iter(ws.members().map(|m| (m.name().as_str(), m)));
+    let base_tree = base_commit.as_object().peel_to_tree()?;
+    let head_tree = head.as_object().peel_to_tree()?;
+    let diff = repo
+        .diff_tree_to_tree(Some(&base_tree), Some(&head_tree), Default::default())
+        .unwrap();
+
+    let mut changed_members = HashMap::new();
+
+    for delta in diff.deltas() {
+        let mut insert_changed = |path: &Path| {
+            if !member_dirs.iter().any(|dir| path.starts_with(dir)) {
+                return;
+            }
+            let Some(member) = path.components().nth(1) else {
+                log::trace!("skipping {path:?}, not a valid member path");
+                return;
+            };
+            let name = member.as_os_str().to_str().unwrap();
+            let Some((name, pkg)) = ws_members.get_key_value(name) else {
+                log::trace!("skipping {name:?}, not found in workspace");
+                return;
+            };
+            if pkg.publish() == &Some(vec![]) {
+                log::trace!("skipping {name}, `publish = false`");
+                return;
+            }
+            changed_members.insert(*name, *pkg);
+        };
+
+        insert_changed(delta.old_file().path().unwrap());
+        insert_changed(delta.new_file().path().unwrap());
+    }
+
+    Ok(changed_members)
+}
+
+/// Compares version against published crates on crates.io.
+///
+/// Assumption: We always release a version larger than all existing versions.
+fn check_crates_io<'a>(
+    config: &Config,
+    changed_members: &HashMap<&str, &'a Package>,
+    needs_bump: &mut Vec<&'a Package>,
+) -> CargoResult<()> {
+    let source_id = SourceId::crates_io(config)?;
+    let mut registry = PackageRegistry::new(config)?;
+    let _lock = config.acquire_package_cache_lock()?;
+    registry.lock_patches();
+    config.shell().status(
+        "BumpCheck",
+        format_args!("compare against `{}`", source_id.display_registry_name()),
+    )?;
+    for member in changed_members.values() {
+        let name = member.name();
+        let current = member.version();
+        let version_req = format!("<={current}");
+        let query = Dependency::parse(name, Some(&version_req), source_id)?;
+        let possibilities = loop {
+            // Exact to avoid returning all for path/git
+            match registry.query_vec(&query, QueryKind::Exact) {
+                task::Poll::Ready(res) => {
+                    break res?;
+                }
+                task::Poll::Pending => registry.block_until_ready()?,
+            }
+        };
+        let max_version = possibilities.iter().map(|s| s.version()).max();
+        if max_version >= Some(current) {
```

Comment:
So we are only finding versions that are `<=` but we need a bump if the version is` >=`?

Shouldn't the version req be `>=` and then we bump if a match is found?